arangod/RocksDBEngine/RocksDBIndex.cpp

```diff
@@ -240,24 +243,83 @@ void RocksDBIndex::afterTruncate(TRI_voc_tick_t) {
   }
 }
 
+
+namespace {
+  bool attributesEqual(VPackSlice first, VPackSlice second,
+                       std::vector<arangodb::basics::AttributeName>::const_iterator begin,
+                       std::vector<arangodb::basics::AttributeName>::const_iterator end) {
+    for (; begin != end; ++begin) {
+      // fetch subattribute
+      first = first.get(begin->name);
+      second = second.get(begin->name);
+      if (first.isExternal()) {
+        first = first.resolveExternal();
+      }
+      if (second.isExternal()) {
+        second = second.resolveExternal();
+      }
+      
+      if (begin->shouldExpand &&
+          first.isArray() && second.isArray()) {
+        auto next = begin + 1;
+        VPackArrayIterator it1(first), it2(second);
+        while (it1.valid() && it2.valid()) {
+          if (!attributesEqual(*it1, *it2, next, end)) {
+            return false;
+          }
+          it1++;
+          it2++;
+        }
+        return true;
+      }
+      
+      int dist = std::distance(begin, end);
+      bool notF1 = first.isNone() || (dist == 1 && !first.isObject());
+      bool notF2 = second.isNone() || (dist == 1 && !second.isObject());
+      if (notF1 != notF2) {
+        return false;
+      }
+      if (notF1 || notF2) { // one of the paths was not found
+        break;
+      }
+    }
+    
+    return (basics::VelocyPackHelper::compare(first, second, true) == 0);
+  }
+} // namespace
+  
+
 Result RocksDBIndex::update(transaction::Methods& trx, RocksDBMethods* mthd,
-                            LocalDocumentId const& oldDocumentId,
-                            velocypack::Slice const& oldDoc, LocalDocumentId const& newDocumentId,
+                            LocalDocumentId const& documentId,
+                            velocypack::Slice const& oldDoc,
                             velocypack::Slice const& newDoc, Index::OperationMode mode) {
   // It is illegal to call this method on the primary index
   // RocksDBPrimaryIndex must override this method accordingly
   TRI_ASSERT(type() != TRI_IDX_TYPE_PRIMARY_INDEX);
+  TRI_ASSERT(oldDoc.isObject() && newDoc.isObject());
+  
+  bool equal = true;
+  for (std::vector<basics::AttributeName> const& path : _fields) {
```

Comment:
how is this check supposed to work with index implementations that do not use '_fields', e.g. IResearchRocksDBLink
looks like for such indexes the updates will never happen
P.S. iresearch internally supports the 'update' operation, it's just not used in ArangoDB because ArangoDB API previously had no such operation on indexes

Reply:
The update operation has been there since 3.3 ?

Reply:
the ArangoDB update operation has always been implemented as a remove()+insert()
there is even no Index::update(...) or MMFilesIndex::update(...)
hence ArangoSearch link implementation just expected a remove()+insert() like all the other indexes